### PR TITLE
Update tests and development of lib

### PR DESCRIPTION
**Pull request by uwaheed88**
- Development of `ckan_auth_api.py`
- Development of `ckan_upload_api.py`
- Development of `ckan_auth_api.py`
- Development of `ckan_client.py` (`ckan_authz()` and `push()` methods)
- Update tests.

### CODE REVIEW

File: lib/ckan_client.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
+from .util import ckan_auth_api
+from .util import ckan_upload_api
+from .file import FileSystem
+
+
+class Uploader:
```
Comment (rufuspollock) [inline lib/ckan_client.py:6]: Do we want a separate Uploader class - would we just call this something like `Client`

File: lib/ckan_client.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
+from .util import ckan_auth_api
```
Comment (rufuspollock) [inline lib/ckan_client.py:1]: I think i'd have the file just called `client.py` after all this is already inside a package called ckan-sdk or similar

File: lib/ckan_client.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+        self.dataset_id = dataset_id
+
+    def ckan_authz(self):
+        # Get the JWT token from CkanAuthz
+
+        scope = 'obj:{}/{}/*:write'.format(self.organization_id, self.dataset_id)
+        response = ckan_auth_api.get_jwt_from_ckan_authz(self.base_url, self.auth_token, scope)
+        return response
+
+    def push(self, file, token, on_progress=None):
```
Comment (rufuspollock) [inline lib/ckan_client.py:31]: Is there a reason you pass token here? It seems a simpler api for (users that) you have:

```
def push(self, file, on_progress)
```

And then this takes care of getting the token.

Also what is file? is it actually a file like object, a path on disk or ...

File: lib/file.py [hunk 4]
```diff
@@ -0,0 +1,37 @@
+import os
+
+import hashlib
+
+
+class FileSystem:
```
Comment (rufuspollock) [inline lib/file.py:6]: I'm dubious you need a class for this but ok ...

it seems a simple method would be sufficient:

```
get_file_info(file_path) => dict
```

**Comment by rufuspollock** [comment]
@uwaheed88 the commit including .pyc files ... is that intentional? (normally i would not expect pyc files in the commit). If you do need to remove them you'll need to submit a new PR (and do new commits) as otherwise we have them in the repo history.

File: lib/util/ckan_auth_api.py [hunk 5]
```diff
@@ -0,0 +1,53 @@
+import logging
+from urllib.parse import urljoin
+
+import requests
+
+log = logging.getLogger(__name__)
+
+def get_jwt_from_ckan_authz(base_url, token, scope):
```
Comment (rufuspollock) [inline lib/util/ckan_auth_api.py:8]: Surely you want this as part of the private methods in your client? That way you don't need to pass `base_url` all over the place.

Comment (rufuspollock) [reply lib/util/ckan_auth_api.py:8]: And this applies for this whole file. I'd imagine these as private methods on your client lib.

File: lib/util/ckan_upload_api.py [hunk 6]
```diff
@@ -0,0 +1,45 @@
+import logging
+import json
+
+import requests
+
+log = logging.getLogger(__name__)
+
+
+def upload_to_storage(url, token, file_content, on_progress):
```
Comment (rufuspollock) [inline lib/util/ckan_upload_api.py:9]: Again - i don't think we need a separate module for this - either private (or public) methods on client or a single util / helper module would suffice IMO.

File: lib/util/ckan_upload_api.py [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+            'oid': file.get_sha256(),
+            'size': file.get_size(),
+        })
+    headers = {
+        'Accept': 'application/vnd.git-lfs+json',
+        'Content-Type': 'application/vnd.git-lfs+json',
+        'Authorization': token,
+    }
+
+    response = requests.post(url, body, headers)
```
Comment (rufuspollock) [inline lib/util/ckan_upload_api.py:40]: This boiler plate would normally get wrapped in a common convenience method e.g.

```
def _make_ckan_request(body, headers) => response
```
```

File: lib/util/ckan_auth_api.py (hunk 5 above)
Comment (rufuspollock) [reply lib/util/ckan_auth_api.py:8]: Is the `token` here the CKAN API key?

Did you consider using the base `ckanapi` library here (if so what were the pros/cons)?

File: tests/test_uploader.py [hunk 8]
```diff
@@ -40,112 +40,83 @@
...
+              ],
+          }
+
+@mock.patch("lib.util.ckan_auth_api.get_jwt_from_ckan_authz", return_value=get_jwt_from_ckan_authz_json)
+@mock.patch("lib.util.ckan_auth_api.request_file_upload_actions", return_value=request_file_upload_actions_json)
+@mock.patch("lib.util.ckan_upload_api.upload_to_storage", return_value=True)
+@mock.patch("lib.util.ckan_upload_api.verify_upload", return_value=True)
+def test_push_works_with_packaged_dataset(get_jwt_from_ckan_authz_mock, request_file_upload_actions_mock,
+                                          upload_to_storage_mock, verify_upload_mock):
+    import pdb; pdb.set_trace()
```
Comment (rufuspollock) [inline tests/test_uploader.py:110]: Committing code with pdb debugging in it is unusual ...

**Comment by rufuspollock** [comment]
@uwaheed88 also the tests are still failing ... 😬

**Review by rufuspollock — CHANGES_REQUESTED** [review]
@uwaheed88 hi usman, i've left a few comments which are more for clarification. Main item is the pyc files and that the tests are not passing ...

File: tests/test_uploader.py (hunk 8 above)
Comment (uwaheed88) [reply tests/test_uploader.py:110]: Sorry about this(pdb) and pyc files!

**Pull request closed without merging**